Newton merge 5: adding the keystone to get access to the Newton pipeline

MFraters opened the pull request:
This pull request can be concidered to be the keystone of the Newton solver in ASPECT. It adds a nonlinear solver to the core and solver so that the Newton solver can actually be used. It is still in a WIP state (I currently only know that it compiles), but the structure should not change too much any more. 

Will look further into this tomorrow.

gassmoeller commented on include/aspect/simulator.h:
```diff
@@ -1223,6 +1226,13 @@ namespace aspect
+       * if we do not need the machinery for doing Newton stuff, we do
+       * not even allocate it.
+       */
+      std_cxx11::shared_ptr<NewtonHandler<dim> > newton_handler;
```
Should this not be an unique pointer?

MFraters replied:
well, only if the MeltHandler should also be a unique pointer I guess

bangerth replied:
All of ASPECT right now uses `shared_ptr` -- we didn't have `std_cxx11::unique_ptr` when we started using these pointers. There will have to be a point where we replace many of these by `unique_ptr`.

MFraters replied:
ok, I will replace it then by unique_prt :)

gassmoeller replied:
remember to use `std_cxx11::unique_ptr` otherwise it will not compile for old compilers

gassmoeller commented on include/aspect/simulator_access.h:
```diff
@@ -607,6 +609,12 @@ namespace aspect
       get_melt_handler () const;
 
       /**
+       * Return a pointer to the newton handler.
```
a reference

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -905,12 +910,22 @@ namespace aspect
-                                                    mask);
+          if (!assemble_newton_stokes_system || (assemble_newton_stokes_system && nonlinear_iteration == 0))
+            {
+              VectorTools::interpolate_boundary_values (dof_handler,
```
please add the `mapping` parameter that is here in master

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -905,12 +910,22 @@ namespace aspect
+            }
+          else
+            {
+              VectorTools::interpolate_boundary_values (dof_handler,
```
same here

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2241,6 +2256,297 @@ namespace aspect
 
+        case NonlinearSolver::Newton_Stokes:
+        {
+          if (parameters.free_surface_enabled)
```
You checked that this never happens above. So you might as well remove it here.

MFraters replied:
No, I made this one of the allowed models for the free surface.

MFraters replied:
I updated the assert message

MFraters replied:
I checked it with the crustal 2d cookbook in the sense that it converges (for the first few timesteps) and creates a mountain as you would expect. I have convergence problems after the first few timesteps (after about 12 timesteps) and I noticed that the iterated Stokes was sometimes outperforming the defect corrected Picard significantly, but I don't think this has anything to do with the free surface, because removing the free surface and leaving the top surface open yields a similar result. It might just be that the Drucker Prager is very difficult, or there is something wrong with an open top boundary, although in the Spiegelman benchmark which is very similar in the setup does not show this behaviour (even the opposite).

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2241,6 +2256,297 @@ namespace aspect
+          double initial_stokes_residual      = 0;
+          std::vector<double> initial_composition_residual (parameters.n_compositional_fields,0);
+
+          //parameters.nonlinear_iteration_number = 0;
```
What about this line? I think your scheme needs to set the nonlinear iteration number somehow. Please take a look how the other schemes handle this.

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2241,6 +2256,297 @@ namespace aspect
+
+          //parameters.nonlinear_iteration_number = 0;
+
+          double max = std::numeric_limits<double>::max();
```
What is max? Please give it a more descriptive name

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2241,6 +2256,297 @@ namespace aspect
+          parameters.newton_residual = 1;
+          parameters.newton_theta = 0;
+
+          const double piccard_linear_stokes_solver_tolerance = parameters.linear_stokes_solver_tolerance;
```
I have the feeling this variable name is as long as the original one, just use the original one.

bangerth replied:
also, the spelling is "picard" (one 'c') ;-)

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2241,6 +2256,297 @@ namespace aspect
+
+          const double piccard_linear_stokes_solver_tolerance = parameters.linear_stokes_solver_tolerance;
+
+          bool   use_picard = true;
```
remove the additional spaces

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2241,6 +2256,297 @@ namespace aspect
+            {
+              assemble_newton_stokes_system = true;
+
+
```
remove the additional empty lines

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2241,6 +2256,297 @@ namespace aspect
+
+          for (unsigned int i=0; (! ((i >= parameters.max_nonlinear_iterations) // regular timestep
+                                     ||
+                                     ((pre_refinement_step < parameters.initial_adaptive_refinement) // pre-refinement
```
You can move a lot of this outside of the for-loop like following:
```
const unsigned int max_iterations = (pre_refinement_step < parameters.initial_adaptive_refinement)
?
std::min (parameters.max_nonlinear_iterations,parameters.max_nonlinear_iterations_in_prerefinement)
:
parameters.max_nonlinear_iterations;

for (unsigned int i=0; i<max_iterations;++i)
...
```
Please check that the logic actually does what it is supposed to do, but if that works it would be much easier to understand than the current form.

bangerth replied:
also, I think `i` is not a good variable name. We really only use it in loops with no more than 3 lines where it is a counting index. But here, it is something like a `nonlinear_iteration` index -- so give it a name that's indicative of what's happening.

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2241,6 +2256,297 @@ namespace aspect
+              /**
+               * Eisenstat Walker method for determining the tolerance
+               */
+              if (nonlinear_iteration > 1)
```
Do you initialize this variable somewhere? And do you increase it during the solver scheme? A quick search did not find anything, but maybe I am missing something.

MFraters replied:
It used to be increased a the end of the loop, but I now made it explicitly part of the loop condition.

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2241,6 +2256,297 @@ namespace aspect
+
+              current_linearization_point.block(introspection.block_indices.temperature)
+                = solution.block(introspection.block_indices.temperature);
+              rebuild_stokes_matrix = true;
```
This line is a bit weird here. Could you move that to the part that deals with the Stokes solver.

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2241,6 +2256,297 @@ namespace aspect
+              current_linearization_point.block(introspection.block_indices.temperature)
+                = solution.block(introspection.block_indices.temperature);
+              rebuild_stokes_matrix = true;
+              std::vector<double> composition_residual (parameters.n_compositional_fields,0);
```
0.0

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2241,6 +2256,297 @@ namespace aspect
+                                         nonlinear_iteration >= parameters.max_pre_newton_nonlinear_iterations))
+                {
+                  use_picard = false;
+                  pcout << "Switching Newton derivatives on." << std::endl;
```
Maybe something like: "Switching to Newton solver scheme"

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2241,6 +2256,297 @@ namespace aspect
+                  use_picard = false;
+                  pcout << "Switching Newton derivatives on." << std::endl;
+
+                  bool use_said_method = false;
```
What is `said`?

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2241,6 +2256,297 @@ namespace aspect
+                                        current_linearization_point,
+                                        global_backup_linearization_point);
+
+                  compute_current_constraints ();
```
Are those two calls necessary before setting the assemblers? It would look more reasonable, if they were placed right before the call where they are needed (I guess inside of `assemble_stokes_system`?)

MFraters replied:
The constraints also need to be computed with the denormalized pressure if I remember it correctly

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2241,6 +2256,297 @@ namespace aspect
+              // need to update the Stokes preconditioner.
+              rebuild_stokes_matrix = rebuild_stokes_preconditioner = assemble_newton_stokes_matrix = true;
+
+              assemble_stokes_system();
```
This means the Stokes system is assembled twice in the first nonlinear iteration (see line 2366)? And this time you do not normalize the pressure afterwards? I have no idea if this is correct or wrong, just want to hear from you that this is what it is supposed to do :smile:

MFraters replied:
Yes, the assembly twice is actually what is supposed to happen. The first is actually only to compute the newton rhs, so I guess I should be able to set assemble_newton_stokes_matrix to false at the beginning.

MFraters replied:
There is actually a normalize pressure, just before the end of the loop :)

MFraters replied:
Why is this unnecessary? This is the only place where the whole matrix is assembled.

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2241,6 +2256,297 @@ namespace aspect
+                        {
+                          if (0.9*parameters.linear_stokes_solver_tolerance * parameters.linear_stokes_solver_tolerance <= 0.1)
+                            {
+                              parameters.linear_stokes_solver_tolerance =  std::min(parameters.minimum_linear_stokes_solver_tolerance, 0.9 * std::fabs(newton_residual*newton_residual)/(newton_residual_old*newton_residual_old));
```
please break this line

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2241,6 +2256,297 @@ namespace aspect
+                                                                                    std::max(0.9 * std::fabs(newton_residual*newton_residual)/(newton_residual_old*newton_residual_old),0.9*parameters.linear_stokes_solver_tolerance*parameters.linear_stokes_solver_tolerance));
+                            }
+                        }
+                      pcout << std::endl << "   The linear solver tolerance is set to " << parameters.linear_stokes_solver_tolerance << ":" << std::fabs(newton_residual-stokes_residual)/newton_residual_old << "," << stokes_residual << "," << newton_residual << ", " << newton_residual_old << std::endl;
```
Please break this line several times :smile:

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2241,6 +2256,297 @@ namespace aspect
+                                pcout << std::endl << "   Solve failed, set Use Picard to true. and the linear solver tolerance is set to "
+                                  << parameters.linear_stokes_solver_tolerance << std::endl;
+                                stokes_residual = solve_stokes();
+                            }*/
```
Remove it, and leave this for a separate pull request, we will also need a test that it works.

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2241,6 +2256,297 @@ namespace aspect
+              double test_newton_residual_velo = 0;
+              double test_newton_residual_pres = 0;
+              double lambda = 1;
+              double alfa = 1e-4;
```
alpha

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2241,6 +2256,297 @@ namespace aspect
+              double alfa = 1e-4;
+              unsigned int line_search_iteration = 0;
+
+              do
```
Please add a comment here, what this loop does. I have lost track by now in which part of the solver we are.

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2241,6 +2256,297 @@ namespace aspect
+                  test_newton_residual_pres = system_rhs.block(introspection.block_indices.pressure).l2_norm();
+                  test_newton_residual = std::sqrt(test_newton_residual_velo * test_newton_residual_velo + test_newton_residual_pres * test_newton_residual_pres);
+
+                  if (test_newton_residual < (1.0 - alfa * lambda) * newton_residual || line_search_iteration >= parameters.max_newton_line_search_iterations || use_picard)
```
break this line for readability

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2241,6 +2256,297 @@ namespace aspect
+
+                  if (test_newton_residual < (1.0 - alfa * lambda) * newton_residual || line_search_iteration >= parameters.max_newton_line_search_iterations || use_picard)
+                    {
+                      pcout << newton_residual/initial_newton_residual << "   Norm of rhs = " << newton_residual << ":" << test_newton_residual  << ", relative residual: " << newton_residual/initial_newton_residual << ", theta = " << parameters.newton_theta << std::endl;
```
break this line into several lines

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2241,6 +2256,297 @@ namespace aspect
+                    {
+
+                      pcout << "   Iteration " << line_search_iteration << ", with norm of the rhs " << test_newton_residual << " and going to "
+                            << (1.0 - alfa * lambda) * newton_residual << ", relative residual: " << newton_residual/initial_newton_residual << ", theta = " << parameters.newton_theta << std::endl;
```
break this one further

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2241,6 +2256,297 @@ namespace aspect
+                      pcout << "   Iteration " << line_search_iteration << ", with norm of the rhs " << test_newton_residual << " and going to "
+                            << (1.0 - alfa * lambda) * newton_residual << ", relative residual: " << newton_residual/initial_newton_residual << ", theta = " << parameters.newton_theta << std::endl;
+
+                      lambda = lambda * (2.0/3.0);// TODO: make a parameter out of this.
```
you can use `*= 2.0/3.0`, also please add a comment on what you are doing here. I suppose you are tightening some kind of tolerance?

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2241,6 +2256,297 @@ namespace aspect
+
+                  line_search_iteration++;
+                }
+              while (true);
```
Every time I see a `while (true)` statement my heart skips a beat :smirk:, are you sure you never get into an endless-loop here, even if the Newton solver does not converge?

MFraters replied:
haha, I changed it, with a comment that in real runs it should actually never become false, because the break statement above should have caught it. Or should I make an assert?

bangerth replied:
Without having looked at the actual code in question, I'm a huge fan of making things explicit: if you think that at some point in code, a particular condition should be true, then put that into an assertion. This way, you can let the code check that indeed everything is as you expect it to be, and it also serves as documentation for others reading the code later on to understand what you thought was the case here.

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2241,6 +2256,297 @@ namespace aspect
+              else
+                {
+                  bool use_said_method = false;
+                  if (use_said_method) //said's method
```
This if is pretty useless right now. Please remove, until you actually use it.

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2241,6 +2256,297 @@ namespace aspect
+
+              pcout << std::endl;
+              this->last_pressure_normalization_adjustment = normalize_pressure(current_linearization_point);
+              ++nonlinear_iteration;
```
Oh, here you increment it. But do you initialize it?

MFraters replied:
it is now part of the main nonlinear loop :)

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2241,6 +2256,297 @@ namespace aspect
+                break;
+            }
+
+          pcout << "Final residual: Norm of rhs = " << newton_residual << ", relative residual: " << newton_residual/initial_newton_residual <<  std::endl;
```
Please adjust the style of this output to be similar to the other solver schemes.

MFraters replied:
Others don't seem to use a final residual print, so I removed it for now

bangerth commented on source/simulator/solver.cc:
```diff
@@ -659,12 +659,20 @@ namespace aspect
+        else
+          {
+            linearized_stokes_initial_guess.block (block_vel) = 0;
+            linearized_stokes_initial_guess.block (block_p) = 0;
```
can you add a comment here that says something like "the Newton solver solves for *updates* to variables, for which we have no good initial guesses. therefore, initialize the solution vector for the solver to zero."?

bangerth commented on source/simulator/solver.cc:
```diff
@@ -861,7 +869,8 @@ namespace aspect
     remove_nullspace(solution, distributed_stokes_solution);
-    this->last_pressure_normalization_adjustment = normalize_pressure(solution);
+    if (parameters.nonlinear_solver != NonlinearSolver::Newton_Stokes)
+      this->last_pressure_normalization_adjustment = normalize_pressure(solution);
```
but do you do the normalization at the end of *all* Newton iterations? this call must be somewhere!

MFraters replied:
yes, it is now in core.cc at the end of every nonlinear iteration

bangerth commented on include/aspect/simulator.h:
```diff
@@ -1418,6 +1428,8 @@ namespace aspect
 
       bool                                                      rebuild_stokes_matrix;
+      bool                                                      assemble_newton_stokes_matrix;
+      bool                                                      assemble_newton_stokes_system;
```
Why do you need these two variables? Are they ever `false`?

bangerth commented on source/simulator/core.cc:
```diff
@@ -2241,6 +2256,297 @@ namespace aspect
+              solution.block(introspection.block_indices.pressure) = 0;
+              solution.block(introspection.block_indices.velocities) = 0;
+
+              pcout << "Newton iteration number: " << nonlinear_iteration << std::endl;
```
I think you should indent that so that it aligns with the rest of the output.

bangerth replied:
also, I don't really understand: your Newton iteration only runs around the Stokes system below, but after this output, you first solve the composition and temperature equation -- which aren't iterated out in the Newton solver. Or are they?

MFraters replied:
I am not sure what you mean. Composition and temperature are solved every iteration (it is analogue to iterated IMPES).

bangerth replied:
Oh, I see -- you mean you solve one Newton step for Stokes, then temperature then composition (or in some other order), and then you start over again? That is interesting. Is this better than just iterating out the Stokes system, then do one temperature and one composition solve?

MFraters replied:
Usually it is fine to do a iterated stokes type of solve (solve temp, solve comp, iterate Stokes) even when your material is nonlinear dependent on temperature and composition. But I have noticed sometimes that the iterated IMPES responds differently. I have based this Newton solver on iterated IMPES because it I think it is technically the best, but iterated Stokes might be faster.

bangerth replied:
I trust your judgment. I'm just thinking of the paper (which I'm working on again today) where we really just do Stokes. I don't know whether it's worth doing iterated Stokes with Newton as well. I guess it's useful to just get the current case you're implementing right, and then doing iterated Stokes with Newton should be a relatively simple addition.

MFraters replied:
ok, I will leave it then for now. I think it is good to have both anyway, but I remember that (I think) @tjhei was working on rewriting the assembler in a way that people can decide from the input file if and when to solve composition, temperature and Stokes.

bangerth commented on source/simulator/core.cc:
```diff
@@ -2241,6 +2256,297 @@ namespace aspect
+                  pcout << "Switching Newton derivatives on." << std::endl;
+
+                  bool use_said_method = false;
+                  if (!use_said_method)
```
that's also a bit funny -- you're setting the variable to `false`, and then you're testing it in the next line in an `if` statement :-)

MFraters replied:
I should make this into a global variable, or remove it. My current preference it making it in a global variable, because I suspect that is could stabilize hard computations, and I would like to be able to quickly test that when other methods fail.

bangerth commented on source/simulator/core.cc:
```diff
@@ -2241,6 +2256,297 @@ namespace aspect
+                    parameters.newton_residual = 0;
+                }
+
+              parameters.newton_theta = std::max(0.,(1-(parameters.newton_residual/parameters.switch_initial_newton_residual)));
```
I think this should be get a better name than `theta` -- nobody is going to know what it is.

MFraters replied:
it is now called newton_derivative_scaling_factor :)

gassmoeller commented:
/run-tests

MFraters commented:
I think that I have addressed all comments, so it is ready for a second round of reviews :)

MFraters commented:
with the merge of #1691 I have been able to actually test the newton solver. It required some fixes in some places, like that I can't compute the velocity_divergence with the help of the trace because it isn't accurate enough. I have now also added two tests of which one is actually using the newton solver. I think it is ready for a final review.

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -532,8 +534,9 @@ namespace aspect
-                     ExcMessage("The free surface scheme is only implemented for the IMPES or Iterated Stokes solver") );
+                     parameters.nonlinear_solver == NonlinearSolver::iterated_Stokes ||
+                     parameters.nonlinear_solver == NonlinearSolver::Newton_Stokes,
+                     ExcMessage("The free surface scheme is only implemented for the IMPES, Iterated Stokes solver or Newton Stokes.") );
```
Add a "solver" at the end of the sentence.

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2145,6 +2162,305 @@ namespace aspect
+          double initial_stokes_residual      = 0;
+          std::vector<double> initial_composition_residual (parameters.n_compositional_fields,0);
+
+          //parameters.nonlinear_iteration = 0;
```
Still there, please remove :smile:

MFraters replied:
you are right, should be removed now :P

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2147,6 +2162,329 @@ namespace aspect
+              linearized_stokes_initial_guess.block (block_vel) = current_linearization_point.block (block_vel);
+              linearized_stokes_initial_guess.block (block_p) = current_linearization_point.block (block_p);
+
+              if (nonlinear_iteration == 0)
```
I suggest you move everything from line 2282 to 2311 into a new helper function called `initialize_newton_solver` or similar. This solver scheme is way too long to keep it in one place.

gassmoeller replied:
Also this code always happens in the first nonlinear iteration, but should it not happen only in the first *Newton* iteration?

MFraters replied:
no, because it computes the initial newton residual, which basically replaces the initial residual from the other nonlinear solvers.

gassmoeller replied:
But then why call it initial newton residual instead of initial residual? If it is the same thing I would prefer to give it the same name.

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2147,6 +2162,329 @@ namespace aspect
+                                        current_linearization_point);
+
+                  compute_current_constraints ();
+                  assemblers.reset (new internal::Assembly::AssemblerLists<dim>());
```
why is this reset and the call to `set_assembers` in the next line necessary? These function should be called once at the beginning of the model to select the appropriate assemblers. Or is this necessary because of the switch from Picard to Newton? Then this call should only happen when we just switched (i.e. around line 2244)

MFraters replied:
Good point, I think this is a old code from when I actually switched assemblers. Maybe I still need it, so I will test this.

MFraters replied:
It was actually necessary, but I have moved it to a more logical position which should be more efficient now :)

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2147,6 +2162,329 @@ namespace aspect
+                  current_linearization_point.block(introspection.block_indices.velocities) = global_backup_linearization_point.block(introspection.block_indices.velocities);
+
+                  parameters.switch_initial_newton_residual = initial_newton_residual;
+                  parameters.newton_residual = initial_newton_residual;
```
There is a problem here. The `parameters` structure is meant for *input* parameters. It should not be abused to store something else in there (imagine what happens if some other part of the code relies on the fact that `parameters` stores the content of the input parameter file). Other solver schemes also store an initial residual, but they do not store that in the `parameters` structure. Is there a reason you need to do that?

MFraters replied:
The problem here is that I need access to these values during the assembly. I agree that this is an ugly way to do it, so if you have a nice solution I will be happy to change it

gassmoeller replied:
I feel like this is a case for member variables of the `NewtonHandler` class. Since we try to make member variables private, so you will also need functions that are called `set_...()` and `get...()`. It is significantly more code, but I feel it is the proper place to put these variables. Alternatively, you could create a member structure of `Simulator` called `NewtonSolverVariables` which contains all the variables you need. But then you need a new `SimulatorAccess` function that provides access to that structure, so in the end it is equally much code. Not sure what is best, what do you think?

MFraters replied:
I will put it into the NewtonHandler. I have reviewed the current dependencies, I might only need one of the three.

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2147,6 +2162,329 @@ namespace aspect
+                  newton_residual_pres = system_rhs.block(introspection.block_indices.pressure).l2_norm();
+                  newton_residual = std::sqrt(newton_residual_velo * newton_residual_velo + newton_residual_pres * newton_residual_pres);
+
+                  if (!use_picard)
```
Please move lines 2340-2375 into its own helper function as well. Also is the function *newton* specific or would it work with other iterative solver schemes as well? If so, please rename the variables to remove Newton specific names. The new function could have a function head like:
```
double Simulator<dim>::compute_linear_solver_tolerance(const double nonlinear_residual,
const double old_nonlinear_residual,
const double stokes_residual)
```
and the new code here could look like:
```
if (!use_picard)
  {
    parameters.linear_stokes_solver_tolerance = compute_linear_solver_tolerance(...);
    pcout << "The linear solver tolerance is set to ...;
  }
```

MFraters replied:
I actually do not know if it can be used for non-Newton methods. I would expect not, because when I read the paper it is specifically designed for the Newton method and it uses the derivatives in the proof (although you don't explicitly need them in the end). But maybe it is possible to generalise the proof for a defect corrected Picard or even a normal Picard iteration. I remember someone at the hackaton stating that it should be possible/useful. Maybe @bangerth or @tjhei can shed some light on that.

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2147,6 +2162,329 @@ namespace aspect
+                  if (!use_picard)
+                    {
+                      const bool EistenstatWalkerChoiceOne = true;
+                      if (EistenstatWalkerChoiceOne)
```
This part needs some documentation (at least what it does, and where to find a description of the method).
Also I think it is Eisenstat instead of Eistenstat, please check.

MFraters replied:
you are right, it is Eisenstat and not Eistenstat :P

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2147,6 +2162,329 @@ namespace aspect
+                      const bool EistenstatWalkerChoiceOne = true;
+                      if (EistenstatWalkerChoiceOne)
+                        {
+                          const double powerterm = (1+std::sqrt(5))*0.5;
```
This magic number needs an explanation (if only something like: This is what they chose in their paper.)

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2147,6 +2162,329 @@ namespace aspect
+              newton_residual_pres = system_rhs.block(introspection.block_indices.pressure).l2_norm();
+              newton_residual = std::sqrt(newton_residual_velo * newton_residual_velo + newton_residual_pres * newton_residual_pres);
+
+              // We may need to do a line search if the solution doesn't decrease the norm of the rhs. TODO: more documentation.
```
see TODO. In particular: Why do you need a backup copy of the current_linearization_point for the line search?

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2147,6 +2162,329 @@ namespace aspect
+                   * This method allows to slowly introduce the derivatives based
+                   * on the improvement of the residual.
+                   */
+                  if (parameters.use_newton_residual_scaling_method) //said's method
```
Does this comment mean: *Said*'s method? Please move it into the longer comment above and mention where to find this method.

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -2147,6 +2162,329 @@ namespace aspect
+                break;
+            }
+
+          // When we are ready with iterating, we need to set the final solution to the current linearlization point,
```
I think in this case one would write: "When we are finished iterating, ..."

gassmoeller commented on source/simulator/parameters.cc:
```diff
@@ -338,6 +338,18 @@ namespace aspect
+                           "of the residual. If we do not use it, we just set it so the newton_derivative_scaling_factor "
+                           "goes from zero to one when switching on the Newton solver.");
+
+        prm.declare_entry ("Minimum linear Stokes solver tolerance", "0.9",
```
I feel like this needs to be called `Maximum linear Stokes solver tolerance` since technically the value is not allowed to be higher. I agree however that using `Maximum` and `Minimum` for solver tolerances is confusing.

gassmoeller commented on tests/nonlinear_channel_flow_Newton_Stokes.prm:
```diff
@@ -0,0 +1,141 @@
+#Like the poiseuille_2d.prm test, but instead of prescribing the
```
I think you need to update the comment at the top of this file.

gassmoeller commented on tests/nonlinear_channel_flow_Newton_Stokes/screen-output:
```diff
@@ -0,0 +1,171 @@
+      Total relative residual after nonlinear iteration 24: 1.33549e-08, norm of the rhs: 9750.95, newton_derivative_scaling_factor: 1
+
+   Skipping temperature solve because RHS is zero.
+   The linear solver tolerance is set to 0.1
```
It seems the linear solver tolerance for this model is always set to the maximum value. Can you add a test that uses the tolerance scaling, and maybe also the newton derivative scaling method? Maybe just tweaking some parameters in this test would do the trick already. Otherwise we can not be sure the method still works when we are going to restructure the nonlinear solver scheme code (related to #1279).

gassmoeller commented on tests/nonlinear_channel_flow_iterated_IMPES/solution/solution-00000.0000.gnuplot:
```diff
@@ -0,0 +1,1799 @@
+# This file was generated by the deal.II library.
```
I would argue that we do not need the visualization output for both new tests. Please add the statistics output for both tests, and remove the visualization output for iterated IMPES (keep it for the Newton test).

MFraters commented:
It took a bit longer to address the comments because I discovered that the velocity boundary condition wasn't working correctly yet. I fixed that and added the nonlinear channel flow test now also for velocity boundary conditions. I also fixed and improved a lot of smaller things.

gassmoeller commented on source/simulator/helper_functions.cc:
```diff
@@ -1574,6 +1574,88 @@ namespace aspect
+
+  template <int dim>
+  void
+  Simulator<dim>::compute_Eisenstat_Walker_linear_tolerance(const bool EisenstatWalkerChoiceOne,
```
I would prefer if you could make the `linear_stokes_solver_tolerance` parameter a `const double` and return the newly computed value. The call to this function could then look like
```
linear_solver_tolerance = compute_Eisenstat_Walker_linear_tolerance( ..., linear_solver_tolerance, ...);
```
It is good practice to avoid functions that modify their parameters, where possible (it is not always possible). They make the code harder to understand.

gassmoeller commented on source/simulator/newton.cc:
```diff
@@ -54,6 +53,24 @@ namespace aspect
+  NewtonHandler<dim>::
+  set_newton_derivative_scaling_factor(const double set_newton_derivative_scaling_factor)
+  {
+    this->newton_derivative_scaling_factor = set_newton_derivative_scaling_factor;
```
no need for `this`

gassmoeller commented on source/simulator/newton.cc:
```diff
@@ -54,6 +53,24 @@ namespace aspect
+  NewtonHandler<dim>::
+  get_newton_derivative_scaling_factor() const
+  {
+    return this->newton_derivative_scaling_factor;
```
no need for `this`

MFraters commented on source/simulator/core.cc:
```diff
@@ -2147,6 +2167,286 @@ namespace aspect
+                      this->compute_Eisenstat_Walker_linear_tolerance(EisenstatWalkerChoiceOne,
+                                                                      parameters.maximum_linear_stokes_solver_tolerance,
+                                                                      parameters.linear_stokes_solver_tolerance,
+                                                                      stokes_residual,
```
you are right, I moved the initialisation a outside the loop and it has (in the few test I did) significantly improved the convergence! :)

MFraters commented:
I think that I have addressed all the comments. The order of the commits showed here is a bit weird, but the one which is tested is the last commit in my log.

gassmoeller commented:
Very nice. Ready from my point of view. If there are no objections from @bangerth or @tjhei I will merge on Monday. Finally! :tada:

bangerth commented:
On 06/23/2017 11:12 AM, Rene Gassmoeller wrote:
> Very nice. Ready from my point of view. If there are no objections from 
> @bangerth <https://github.com/bangerth> or @tjhei <https://github.com/tjhei> I 
> will merge on Monday. Finally! 🎉

All, would you mind hanging on to this for another week and a half? I would 
like to take a look as well -- if only to familiarize myself with the code, 
but in reality one always finds something. The problem is that I have, on 
average, maybe 10 minutes of internet per hour of being at work. It is not 
enough to look through big patches, unfortunately, but I will make it a 
priority after I get back next weekend.

MFraters commented:
I don't mind. I am currently running all kinds of benchmarks and parameters/performance tests using this branch. I will also add a changes entry to this pull request.

gassmoeller commented:
@bangerth: Do you have time to give this a look this week? It is currently blocking #1785 and #972, which I would like to finish over the summer.

bangerth commented on source/simulator/core.cc:
```diff
@@ -2147,6 +2167,286 @@ namespace aspect
+               */
+              LinearAlgebra::BlockVector backup_linearization_point = current_linearization_point;
+
+              double test_residual = 0;
```
Enclose the whole line search code into a set of braces to make sure that all of these variables that you declare here only live until we're done with the line search. It's not necessary for them to live on after we're done with the line search.

bangerth commented on source/simulator/core.cc:
```diff
@@ -2147,6 +2167,286 @@ namespace aspect
+
+
+                  current_linearization_point.block(introspection.block_indices.pressure) += search_direction.block(introspection.block_indices.pressure);
+                  current_linearization_point.block(introspection.block_indices.velocities) += search_direction.block(introspection.block_indices.velocities);
```
I think you can merge the multiplication of `search_direction` and the addition here by using the `add(double,Vector)` member function of the vector classes. This way, you don't actually have to touch the search direction vector.

MFraters replied:
doing this current_linearization_point.block(introspection.block_indices.velocities).add(search_direction.block(introspection.block_indices.velocities)); doesn't work because the vector has ghost elements.

bangerth commented on source/simulator/core.cc:
```diff
@@ -2147,6 +2167,286 @@ namespace aspect
+                  assemble_newton_stokes_matrix = rebuild_stokes_preconditioner = false;
+                  rebuild_stokes_matrix = true;
+
+                  assemble_stokes_system();
```
Why do you need to rebuild the stokes matrix here? Also, I seem to have forgotten the difference between `assemble_newton_stokes_matrix` and `rebuild_stokes_matrix` -- can you refresh my mind?

MFraters replied:
`assemble_newton_stokes_matrix` determines whether only the newton rhs is build or the whole newton system. 

`rebuild_stokes_matrix` determines whether information for rebuilding the Stokes matrix needs to be (re)computed. For the Newton solver this variable has to be true all the time because even the rhs needs this info.

bangerth commented on source/simulator/core.cc:
```diff
@@ -2147,6 +2167,286 @@ namespace aspect
+                  Assert(line_search_iteration <= parameters.max_newton_line_search_iterations,
+                         ExcMessage ("This tests the while condition. This condition should "
+                                     "actually never be false, because the break statement "
+                                     "above should have caught it."));
```
This is not a useful message to users. And the condition may happen -- you just haven't found a sufficiently small update that satisfies the stopping condition. This can happen for poorly conditioned problems and/or insufficiently solved linear systems.

I think I would just say something like "The line search as part of the Newton method did not find a suitable update to the current solution within the allowed number of steps. Try allowing more steps, solving the linear system more accurately, or reducing the degree of ill-posedness of the original problem."

MFraters replied:
Well, I don't think this assert can every be triggered unless someone changed the code which is about 25 lines above this line (the code where it basically states if line_search_iteration >= parameters.max_newton_line_search_iterations -> break (note that there is a line_search_iteration++; in between). So what I basically say in this assert is that you changed the code just above this assert in such a way that it is broken. So it was meant more for developers than for users. 

I personally think that we should remove this assert, because I doubt  how useful this assert is in reality for developers, because they might be expected to know what they are doing before changing core.cc and test what they changed anyway. Or do you want these kind of securities in the code?

MFraters replied:
And there is also the while statement below, which is also a safeguard. So it is currently trippel safeguarded :P

bangerth commented on source/simulator/core.cc:
```diff
@@ -2147,6 +2167,286 @@ namespace aspect
+                                     "above should have caught it."));
+                }
+              while (line_search_iteration <= parameters.max_newton_line_search_iterations);
+              // The while condition should actually never be false, because the break statement above should have caught it.
```
break the comment

MFraters commented on benchmarks/nonlinear_channel_flow/simple_nonlinear.cc:
```diff
@@ -0,0 +1,423 @@
+     * @ingroup MaterialModels
+     */
+    template <int dim>
+    class SimpleNonlinear : public MaterialModel::Interface<dim>, public ::aspect::SimulatorAccess<dim>
```
I have no strong feelings where it should go. It used to be a separate material model, but Rene suggested that because it was unlikely to be used outside of benchmarking and tests it was better to put it in the benchmark folder, so I moved it there.

gassmoeller replied:
I would prefer it to be a benchmark material model. It has constant properties, except for an only strain-rate dependent viscosity. While obviously useful for benchmarking the Newton solver, I can not imagine a production run that would use this material model.

MFraters commented on include/aspect/newton.h:
```diff
@@ -71,6 +71,22 @@ namespace aspect
+       * Gets the Newton derivative scaling factor used for scaling the
+       * derivative part of the Newton Stokes solver in the assembly.
+       */
+      double get_newton_derivative_scaling_factor() const;
```
Yes, that is correct. I will add that and also add a sentence that setting it to zero results into a defection correction for of the Picard iteration.

MFraters commented on include/aspect/simulator.h:
```diff
@@ -1194,6 +1197,27 @@ namespace aspect
+
+      /**
+       * This function computes the Eisenstat Walker linear tolerance used for the Newton Stokes solver.
+       * The Eisenstat and Walker (1996) method is used for determining the linear tolerance of
```
Yes, it refers to the 1996 paper (Choosing the forcing terms in an inexact Newton method). How do you want me to reference it exactly here? Is just adding the doi (http://dx.doi.org/10.1137/0917003) enough?

MFraters commented on include/aspect/simulator.h:
```diff
@@ -1236,6 +1260,13 @@ namespace aspect
+       * if we do not need the machinery for doing Newton stuff, we do
+       * not even allocate it.
+       */
+      std_cxx11::unique_ptr<NewtonHandler<dim> > newton_handler;
```
Yes, that is the reason. So should I change the other one into a unique pointer or this one in a shared pointer, or just leave it as it is?

MFraters commented on source/simulator/helper_functions.cc:
```diff
@@ -1574,6 +1574,90 @@ namespace aspect
+  {
+    LinearAlgebra::BlockVector temp_linearization_point = current_linearization_point;
+    const unsigned int block_vel = introspection.block_indices.velocities;
+    temp_linearization_point.block(introspection.block_indices.velocities) = 0;
```
haha, good point :)

bangerth commented:
Either way is fine as long as the two places are consistent. Long term, we want to move to unique pointers, so maybe that's the way to go. 
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

MFraters commented:
@gassmoeller One problem which I encounter now the newton_derivative_scaling_factor is no longer in parameters, is that when running tests like the simple_nonlinear.cc test on material models which want to use the newton_derivative_scaling_factor for optimalisation purposes (don't need to compute the derivatives when the newton_derivative_scaling_factor = 0), it causes a very ugly segmentation fault for that test, because the newton handler doesn't exists.

MFraters commented on source/simulator/core.cc:
```diff
@@ -2147,6 +2167,286 @@ namespace aspect
+          double velocity_residual = 0;
+          double pressure_residual = 0;
+          double residual = 1;
+          double residual_old = 1;
```
I have three different residuals, these two (they used to be called Newton_Stokes_residual) which I compute by caluculting the l2 norm of the Newton rhs (F) and an other one which is currently called the stokes residual and which is what the solve_stokes returns (normaly Ax-b, now with Newton Jdx-F). What are good names for these different variables?

MFraters commented:
I have put the changes in a separate commit so that it is hopefully easier to review. There are a several changes I didn't make yet, because they are still in discussion.

gassmoeller commented:
Hi Menno,
sorry for coming back late to your comment from a week ago. Did you solve the issue with the segmentation fault in the simple_nonlinear test? The test seems to run for me.

MFraters commented:
Hey Rene,
Sorry I might not have been clear on that issue. The issue only happens on material models which have been optimalised by using the newton_derivative_scaling_factor (that is when the derivatives are not needed, they are not computed). In this case the simple_nonlinear.cc material model has not been optimalised in this way, so the simple_nonlinear.cc test doesn't give a problem.

But I have been testing other material models which have been optimalised in that way, they they show this problem. I only remembered when I encountered this problem again, that that was the reason I had put the variable newton_derivative_scaling_factor in parameters. Because the parameters are available during the kind of test I do in the simple_nonlinear.cc test.

So, or I will have to put the variable back into parameters, or we need to find an other solution, or we need to fundamentally change the tests of these material models.

gassmoeller commented:
Hmm, I think I see the problem. 
Just to make sure I understand that correctly: You want to optimize some of your material models based on something that happens (and is computed) inside of `Simulator`, but you want to test it without an instance of `Simulator` existing, and that is why you put the variable into `parameters` instead. This seems like a bad idea to me. I would say it is simply impossible to test these optimized models without a real `Simulator` around it. But then I do not know how complicated that would be, so consider my opinion an outsiders opinion here.

To be clear: I am fine with the current version of this PR, and would prefer to not make the structure less intuitive just to make testing simpler.

MFraters commented:
Ok, let's leave it as it is. I will probably just add an extra parameter in the material model itself to switch the optimalisation on and off. I will add the logfile as soon as the open discussions with @bangerth have been finished.

MFraters commented:
I have in the mean time performed many tests and reduced the calls to the compute_current_contraints function, so it is only called for the first two iterations. I noticed that before this change it took up more then 60 percent of the total run time on large runs distributed over 10 nodes and now only about 4 precent :)

But there is an other problem, which I think is important I am a bit confused about. In assemblers/newton_stokes.cc at line 255 there is a debug code which is supposed to check whether the local matrix is SPD. 

```
#if DEBUG
              // Testing whether the Jacobian is Symmetric Positive Definite (SPD)
              if (assemble_newton_stokes_matrix)
                {
                  bool testing = true;
                  for (unsigned int sample = 0; sample < 10; ++sample)
                    {
                      Vector<double> tmp (stokes_dofs_per_cell);

                      for (unsigned int i=0; i<stokes_dofs_per_cell; ++i)
                        if (scratch.finite_element_values.get_fe().system_to_component_index(i).first < dim)
                          tmp[i] = Utilities::generate_normal_random_number (0, 1);

                      const double abc =  data.local_matrix.matrix_norm_square(tmp)/(tmp*tmp);
                      if (abc < -1e-12*data.local_matrix.frobenius_norm())
                        {
                          testing = false;
                          std::cout << sample << " Not SPD: " << abc << "; " << std::endl;

                          for (unsigned int i=0; i<stokes_dofs_per_cell; ++i)
                            {
                              for (unsigned int j=0; j<stokes_dofs_per_cell; ++j)
                                std::cout << std::setprecision(1)  << data.local_matrix(i,j) << "," << std::flush;
                              std::cout << "},{" << std::endl;
                            }
                          std::cout << std::endl;
                          std::cout << std::setprecision(6) << std::endl;

                          Assert(testing,ExcMessage ("Error: Assembly not SPD!."));

                          // Testing whether all entries are finite.
                          for (unsigned int i=0; i<stokes_dofs_per_cell; ++i)
                            {
                              for (unsigned int j=0; j<stokes_dofs_per_cell; ++j)
                                {
                                  Assert(dealii::numbers::is_finite(data.local_matrix(i,j)),ExcMessage ("Error: Assembly matrix is not finite."));
                                }
                            }
                        }
                    }
                  if (testing == false)
                    std::cout << std::endl;
                }
#endif
```

The problem I encounter is that is is triggered almost constantly even when I place it a few lines higher at the part where no derivatives are added to the matrix. So or something must be very fundamentally wrong with the newton code or the test if flawed. But when I disable the test, the code just works fine (crashes often without the spdfactor and is very stable with it).

I am afraid that we first need to resolve this problem before we can merge and celebrate :(

tjhei commented:
> I am afraid that we first need to resolve this problem before we can merge and celebrate :(

I think @bangerth should take a look at this. Maybe he has an idea.

bangerth commented:
Hm, that really is weird. Are you saying that this also happens if the matrix you assemble is really just the regular Stokes matrix (i.e., the one with only eta, without the term `d eta/d eps : eps(u)`)? That would make no sense. If that's the case, can you post a patch/diff against your branch with the check above added, and an input file that demonstrates this?

bangerth commented:
Some updates to the branch are here: https://github.com/MFraters/aspect/pull/14

MFraters commented:
@bangerth Yes, I find that very weird as well.  After the changes you made, the problem is still there. I will upload the material model in a separate branch plus an input file to show the problem.

MFraters commented on source/simulator/helper_functions.cc:
```diff
@@ -1500,10 +1500,10 @@ namespace aspect
 
     // we use a different (potentially smaller) time step than in the advection scheme,
     // and we want all of our reaction time steps (within one advection step) to have the same size
-    const int number_of_reaction_steps = std::max((int) (time_step / parameters.reaction_time_step),
```
@bangerth I do not think this code belongs here. Should I remove it from the my pull request?

MFraters commented:
@bangerth sorry, it took a while. I was reminded by Anne and Rene that I already had made the drucker prager so that the derivatives are computed. So I just have to share an input file which you should be able to run and which triggers the spd test even if the derivatives are not added:

```
set Dimension = 2
set CFL number                             = 1.0
set End time                               = 0
set Start time                             = 0
set Adiabatic surface temperature          = 0
set Surface pressure                       = 0
set Use years in output instead of seconds = false  # default: true
set Nonlinear solver scheme = Newton Stokes
set Number of cheap Stokes solver steps = 0
set Max nonlinear iterations = 150
set Nonlinear solver tolerance = 1e-14
set Linear solver tolerance = 1e-5
set Pressure normalization = no

subsection Solver parameters
  subsection Newton solver parameters
        set Max pre-Newton nonlinear iterations = 50
    set Nonlinear Newton solver switch tolerance = 1e-20
        set Max Newton line search iterations = 0
    set Maximum linear Stokes solver tolerance = 9e-1
        set Use Newton residual scaling method = false
  end
end

subsection Boundary temperature model
  set Model name = box
end

 subsection Initial temperature model
   set Model name = function
 end

 subsection Initial composition model
   set Model name = function
   subsection Function
    set Variable names = x,z
    set Function constants = H = 30e3
    set Function expression = if(z < 0.25 * H,1, \
if(z < (0.25+(1/12)-0.02)*H & x > (2-(1/12))*H & x < (2+(1/12))*H,1, \
if(z < (0.25+(1/12))*H & (x > (2-(1/12)+0.02)*H & x < (2+(1/12)-0.02)*H | (x-(2-(1/12)+0.02)*H)^2 + (z-(0.25+(1/12)-0.02)*H)^2 < (0.02*H)^2 | (x-(2+(1/12)-0.02)*H)^2 + (z-(0.25+(1/12)-0.02)*H)^2 < (0.02*H)^2),1, \
if(z < (0.25+0.02)*H && x > (2-(1/12)-0.02)*H && x < (2+(1/12)+0.02)*H && ((x-(2-(1/12)-0.02)*H)^2 + (z-(0.25+0.02)*H)^2 > (0.02*H)^2 && (x-(2+(1/12)+0.02)*H)^2 + (z-(0.25+0.02)*H)^2 > (0.02*H)^2),1,0))))

    #set Function expression = if(z > 1e3 | (x < 19e3 | x > 21e3), 0, 1)
   end
 end

subsection Gravity model
  set Model name = vertical
  subsection Vertical
    set Magnitude = 9.81
  end
end

subsection Geometry model
  set Model name = box

  subsection Box
    set X extent = 120e3
    set Y extent = 30e3
    set X repetitions = 4
  end
end

subsection Material model
    set Model name = drucker prager
end

subsection Mesh refinement
  set Initial adaptive refinement        = 0
  set Initial global refinement          = 5
  set Strategy =  strain rate
end

subsection Model settings
  set Fixed temperature boundary indicators   = 
  # Prescribe a horizontal traction on the vertical boundaries
  set Prescribed traction boundary indicators = 
  set Tangential velocity boundary indicators = 2 
  set Zero velocity boundary indicators       =  
  # Prescribe a zero vertical velocity component on the vertical boundaries
  set Prescribed velocity boundary indicators = 0 x:function,1 x:function
end


subsection Boundary traction model
  subsection Function
    set Variable names = x,z
    # We want to prescribe a pressure of 2 at the left boundary
    # and -2 at the right boundary. 
    # The traction in this case is defined as:
    # tau =  - pressure * normal_vector.
    # On the left boundary, the outward pointing normal vector is 
    # (-1;0). On the right (1;0).
    # Therefore:
    # Left boundary:  tau = - pressure(left) (-1;0) = - (2) (-1;0) = (2;0).
    # Right boundary: tau = - pressure(right) (1;0) = - (-2) (1;0) = (2;0). 
    # Conveniently, the traction is the same on both boundaries.
    set Function expression = 0;1.575e8
  end
end

subsection Compositional fields
  set Number of fields = 1
end

subsection Boundary velocity model
  subsection Function
    set Variable names = x,z
    set Function expression = if(x<60e3,7.92219116e-11,-7.92219116e-11);0
  end
end

subsection Postprocess
  set List of postprocessors = velocity statistics, pressure statistics, mass flux statistics, visualization
  subsection Visualization
    set List of output variables = material properties,strain rate, spd factor
    set Interpolate output = false
  end
end
```

gassmoeller commented:
MFraters#15 hopefully fixes the SPD-test issues. 

And I would not worry about the test changes. The differences in the failing test `nonlinear_channel_flow_tractions_Newton_Stokes` are very minor (likely some round-off error), and the accompanying test `nonlinear_channel_flow_velocities_Newton_Stokes` does not show any changes. Maybe something changed on master during a rebase, or the changes in MFraters#14 resulted in a different order of operations that caused round-off differences.

Could you maybe extract the changes to melt_handler (pointer types) and reaction_steps (int -> unsigned int) into two separate PRs? Those can be merged immediately.

MFraters commented:
Thanks again Rene for looking at this problem! I will update the tests results, split the melt_handler and reaction steps into two other pull requests and finally add a log entry and then hopefully we can finally merge this pull request ;)

MFraters commented:
I had some problems with git rebase, but it looks like everything is fixed now. I sorted out the pieces of code which should be separately  committed (of which one was already and the other is now in #1934), updated the test results and added a change file, and it all passes the tester now!

gassmoeller commented:
178 comments, and now it is done. I am not sure if I am brave enough to press the merge button, after all it is the end of a neverending story. :smile:
Thanks for doing all this and sticking with us through all the comments, I cant wait to see it in action :+1:
